rust/src/array.rs

```diff
@@ -18,447 +18,562 @@
+    fn from(data: ArrayDataRef) -> Self {
+        debug_assert!(data.buffers().len() == 1);
+        debug_assert!(data.child_data().len() == 1);
+        let values = make_array(data.child_data()[0].clone());
+        let value_offsets = data.buffers()[0].raw_data() as *const u8 as *const i32;
+        Self {
+            data: data.clone(),
+            values: values,
+            value_offsets: RawPtrBox::new(value_offsets),
         }
-        n
     }
+}
 
-    /// Determine the maximum value in the array
-    pub fn max(&self) -> Option<T> {
-        let mut n: Option<T> = None;
-        match &self.validity_bitmap {
-            &Some(ref bitmap) => for i in 0..self.len {
-                if bitmap.is_set(i) {
-                    let mut m = self.data.get(i);
-                    match n {
-                        None => n = Some(*m),
-                        Some(nn) => if *m > nn {
-                            n = Some(*m)
-                        },
-                    }
-                }
-            },
-            &None => for i in 0..self.len {
-                let mut m = self.data.get(i);
-                match n {
-                    None => n = Some(*m),
-                    Some(nn) => if *m > nn {
-                        n = Some(*m)
-                    },
-                }
-            },
-        }
-        n
+impl Array for ListArray {
+    fn as_any(&self) -> &Any {
+        self
+    }
+
+    fn data(&self) -> ArrayDataRef {
+        self.data.clone()
+    }
+
+    fn data_ref(&self) -> &ArrayDataRef {
+        &self.data
     }
 }
 
-/// Implement the Add operation for types that support Add
-impl<T> PrimitiveArray<T>
-where
-    T: ArrowPrimitiveType + Add<Output = T>,
-{
-    pub fn add(&self, other: &PrimitiveArray<T>) -> PrimitiveArray<T> {
-        let mut builder: Builder<T> = Builder::new();
-        for i in 0..self.len {
-            let x = *self.data.get(i) + *other.data.get(i);
-            builder.push(x);
+/// A special type of `ListArray` whose elements are binaries.
+pub struct BinaryArray {
+    data: ArrayDataRef,
+    value_offsets: RawPtrBox<i32>,
+    value_data: RawPtrBox<u8>,
+}
+
+impl BinaryArray {
+    /// Returns the element at index `i` as a byte slice.
+    pub fn get_value(&self, mut i: i64) -> &[u8] {
+        i += self.data.offset();
+        unsafe {
+            let pos = self.value_offset_at(i);
+            ::std::slice::from_raw_parts(
+                self.value_data.get().offset(pos as isize),
+                (self.value_offset_at(i + 1) - pos) as usize,
+            )
         }
-        PrimitiveArray::from(builder.finish())
     }
-}
 
-impl<T> Array for PrimitiveArray<T>
-where
-    T: ArrowPrimitiveType,
-{
-    fn len(&self) -> usize {
-        self.len
+    /// Returns the element at index `i` as a string.
+    pub fn get_string(&self, i: i64) -> String {
+        let slice = self.get_value(i);
+        unsafe { String::from_utf8_unchecked(Vec::from(slice)) }
```

Comment:
For the record, does `Vec::from(slice)` make a copy?

Reply:
Yes, it will copy the `slice`.